### DISCUSSION TITLE
SOUNDEX matching in product search allowing spelling mistakes

### DISCUSSION THREAD

**Discussion by neekfenwick** [zencart/zencart#6034] [Ideas]
I'm working on a modification to our Zen Cart 1.5.8 store that allows customers to make spelling mistakes when typing into the Search box and still get results.  Is this something that has been solved before?

I'm approaching it by making zero core modifications, like an addon would.  Wondering if it could be considered a core modification, probably with the option by store admin to en/disable it.

My approach is to prepare a lookup database table for each field being matched against, currently I only care about `products_description.products_name` (but I know the search function can match against description, meta keywords & description, etc).  I have a stored procedure that splits each name into individual words and stores a row per word with its soundex value in the `products_name_soundex` table.  The search function then builds `WHERE SOUNDEX(keyword) = products_name_soundex.word_soundex` instead of `WHERE products_name LIKE '%keyword%'`.  We need a per-word lookup table so that individual words can be matched, we cannot simply say `WHERE SOUNDEX(products_name) = SOUNDEX(keyword)`.

If it were broadened to support other tables and fields I imagine the lookup table would be expanded by adding columns specifying the table and fields, rather than adding a lookup table per table/field combination, but I haven't really thought that far ahead.

I'm modifying the search SQL using some quite clumsy text search and replace in an Observer, since the latest core modifications to `search_result/header.php` pass the SQL being built into notifiers.  So it replaces "pd.products_name LIKE '%foo%'" using a regex.  If approved, it could do with a critical eye from impartial devs, as I tend to work in a bit of a silo.

Cons:
- does not support substring search, as the `LIKE '%keyword%'` matching does.
- SOUNDEX is not a particularly good algorithm, but it's built in and good enough so far for my purposes.  From https://dev.mysql.com/doc/refman/8.0/en/string-functions.html#function_soundex "This function implements the original Soundex algorithm, not the more popular enhanced version (also described by D. Knuth). The difference is that original version discards vowels first and duplicates second, whereas the enhanced version discards duplicates first and vowels second."
- SOUNDEX favours the English language, I haven't dug into a good solution in other languages or character sets.

SOUNDEX ought to be reliable on ZC installs in the wild, from https://docs.zen-cart.com/user/first_steps/server_requirements/ the DB requirements are stated as "MySQL 5.7.8+ or MariaDB 10.2.7+".

I'm now working on a trigger on `products_description` table so that updates to the table re-generate the cooked soundex values in the lookup table, so they stay up to date.  Alternatively, the admin functions that update a product could manually call the procedure but I'd rather keep this at the database trigger level to support data loading scripts and whatnot.

Properly indexed, the explain plan shows the matching to be very efficient.

**Comment by drbyte** [comment]
> So it replaces `"pd.products_name LIKE '%foo%'"` using a regex.

Yes, that's probably the simplest way. 

> does not support substring search, as the `LIKE '%keyword%'` matching does

I suppose if you wanted to keep the substring/wildcard matching capability you could `OR` it by keeping the LIKE logic instead of replacing it. ie: `(LIKE-clause OR SOUNDEX-clause)` 

I'm assuming you're using this observer hook:
`$zco_notifier->notify('NOTIFY_SEARCH_WHERE_STRING', $keywords, $where_str, $keyword_search_fields);`

**Reply by neekfenwick** [reply]
Good idea on the OR.   Yes, that and another hook, my observer is very simple right now:

```
    switch ($notifier) {
      case 'NOTIFY_SEARCH_FROM_STRING':
        // Add the join for the soundex lookup table to the FROM clause
        $param2 .= "\nLEFT OUTER JOIN products_name_soundex pns ON pns.products_id = p.products_id";
        break;
      case 'NOTIFY_SEARCH_WHERE_STRING':
        // Modify the built WHERE clause so the fields we know are mapped in the lookup table are matched
        // Look for eg. pd.products_name LIKE '%anything%'
        // Replace with pns.word_soundex = SOUNDEX('anything')
        $param2 = preg_replace("/pd\.products_name LIKE '([^']*)'/", "pns.word_soundex = SOUNDEX('\\1')", $param2);
        break;
    }
```
I've not put much thought into non-alphabetic characters, mostly just trying to placate our fat fingered customers, and we have some quite hard to spell product names like 'anthracite' and 'azurite'.

I'd also like to split the function that does the search into a helper, refactoring it out of `search_results/header.php`, so that I can call it from an `ajax.php` handler, as it would be nice to provide a dropdown on the search input field where the user can select from the matched products without having to visit the results page.  I would solve this on the UI with the Dojo framework that we employ extensively, ~~but I don't think the catalog side has jQuery by default so this wouldn't fit a core  #modification~~ *edit* correction, I have hacked jQuery out of our site's frontend because it did horrible things to our SEO, but if it's there in core then perhaps this would work, I'll have to look at how dropdowns work in jQuery-ui.

**Reply by drbyte** [reply]
Have you looked at any of the instant-search plugins? IIRC one of them hooks the ajax handler. But probably has its own bespoke search script cloned from the search-results-headerphp. Might find some idea-inspiration of ways to tackle it.

A class to contain that search logic would probably be ideal. Could probably move functions_search functions into it too.
The admin has even more search functions, but implemented differently for its own reasons. I'm sure there's some overlap.

**Reply by neekfenwick** [reply]
@drbyte do you think this feature is likely to make it into core, in which case I can integrate changes tightly in the search functions, or not, in which case I need to continue down the addon/observer route?  If it's core, I would have to think more about adding configuration options to turn it off, as I can see some sites might have search terms that are deliberately similar and would match by SOUNDEX but are not intended to be equivalent.  Bear in mind I'm creating a stored procedure and trigger as well as new table, so the installation/upgrade procedure is fairly complex.  I'm up to it, just need to know what direction to head in as I polish it.

**Reply by drbyte** [reply]
As to the SOUNDEX portion, I don't think I'd make it part of core. 

But I've always been an advocate of making reasonable adaptions to core that support such plugins to be easily installed.
So, for example, reworking the search stuff to make it accessible in more places than where it is now (ie: move out of header_php into its own class), and making said class relatively easy to hook into (with observers and if really needed then also constructor params and/or getters/setters to control internal properties), both for "this" plugin's needs, and with a view to how other plugins would work too ..... such an approach would be easily justifiable to put into core. 

As for your configuration options specific to SOUNDEX (or bypassing it in certain cases or on certain phrases), it'd be sensible that those things be part of your plugin.

As for SOUNDEX specifically, as you say, there are pros and cons of that specific technical approach, and while it suits your needs presently, it may need to be swapped out for something else. Architecting it in such a way that core code allows your plugin to be bolted on easily, but perhaps even allowing a swap-out for another engine, would give most long-term flexibility.

That's my current viewpoint anyway.

**Reply by neekfenwick** [reply]
Great, that's a good steer @drbyte.  I'll make a global variable `$search`, instantiated via autoloader and defined in `config.core.php`, with the guts from `search_results/header_php.php` factored out into a helper class.  Work in progress.

When I try to open an Issue for this, with a view to linking a PR, I see "NOTE: Please open feature requests in the Discussions tab, not the Issues tab.  Issues that are actually feature requests will be migrated to Discussions." - can one link a PR to a Discussion?  The docs at https://docs.github.com/en/discussions/quickstart indicate that Discussions are not related to code.  I'll hold off opening an Issue for now.

**Reply by neekfenwick** [reply]
I've issued a PR  https://github.com/zencart/zencart/pull/6036 . I haven't integrated with an existing addon, approached jQuery, or touched the ZC admin side with my changes.

I had a look at Instant Search 4.0 https://github.com/marco-pm/zencart_instantsearch and it's pretty advanced, I wouldn't want to write my own to do SOUNDEX matching when that exists and is so well developed.  At the same time, it has its own SQL building from scratch, so would not benefit from this refactoring, and uses jquery where i'm pretty sure the JS loading is inefficient, so I wouldn't use it on my live site.  So, I think I'll probably just cook my own search dropdown with ajax support, using the search refactoring in my PR.

The ZC admin side also does a lot of manual query building, the only shared code is the function that builds the WHERE clause from keywords, which is also called from my refactoring.  I might try approaching that code at some point but I think the work done so far here is a step in the right direction.

**Comment by neekfenwick** [comment]
I have released Soundex Search addon, see download page https://www.zen-cart.com/downloads.php?do=file&id=2380 and support thread at https://www.zen-cart.com/showthread.php?229817-Soundex-Search-Support-Thread&p=1398022

Closing this discussion.  Thanks for all the input!  Any feedback on the addon would be welcome, probably on the github project at https://github.com/neekfenwick/soundexsearch rather than on the support thread.
